Calypsoify: Refactor and add autoupdate functionality

georgestephanis opened the pull request:
Iterations on the current pass.  Reorganize some assets, tidy up some formatting and code patterns, convert the css to scss to make for easier maintenance going forward and abstract out the color codes to allow for semantic renaming in a future pass.

Oh, and add autoupdate management functionality to the plugins page.

![pageshot of plugins test local wordpress thingy wordpress 2018-11-09-0843 24](https://user-images.githubusercontent.com/941023/48265919-d8e54280-e3fb-11e8-96db-3b789a7a502c.png)

Fixes #10474

#### Changes proposed in this Pull Request:

* Improve Calypso integration for relevant sites, and add admin ui autoupdate functionality.

#### Testing instructions:

* Install some plugins that are not in the repository (you should be able to select some from jurassic.ninja/specialops)
* Open your dev console and keep an eye out for js errors 
* Visit `/wp-admin/plugins.php?calypsoify=1`
* Test out the autoupdate plugin toggles
* Test out the bulk-select of auto-updates 
* Visit `/wp-admin/` to clear calypsoify skin.

georgestephanis committed:
Move Calypsoify class directory down from modules.

georgestephanis committed:
Pull css into scss files, then begin nesting and tidying

georgestephanis committed:
Sass-ize the scss

Build command: `sass style.scss | perl -pe 's/^    /\t\t/g' | perl -pe 's/^  /\t/g' > style.css`

georgestephanis committed:
Should be a space after the colon

georgestephanis committed:
Tidy out and extract more colors.

This will make changes and rebasing easier.  We can also change some colors to be calculated from others.

georgestephanis committed:
Finish swapping out colors for variables

georgestephanis committed:
Extract colors to a common file, build both css files from sass sources.

georgestephanis committed:
Tidy up some jQuery references and how we're setting window.load

georgestephanis committed:
This shouldn't need to be included twice if it is somehow.

georgestephanis committed:
Some minor tidying of the gutenberg js -- swapping var to const

georgestephanis committed:
Pretty up the icons for non-masterbar local mode

georgestephanis committed:
Force the admin ui color to be default `fresh`

georgestephanis committed:
Add visual toggle for enabling automatic updates.

Doesn't do anything yet, that comes next.

georgestephanis committed:
Add ability to manage autoupdates for plugins.

georgestephanis committed:
Add in bulk actions for plugin autoupdates.

Co-Authored-By: Mark Uraine <mapk@users.noreply.github.com>

dereksmart commented on modules/calypsoify/mods-gutenberg.js:
```diff
@@ -1,7 +1,7 @@
 jQuery( function( $ ) {
-	var editPostHeaderInception = setInterval( function() {
-		var $editPostHeader = $( '.edit-post-header' );
+	const editPostHeaderInception = setInterval( function() {
```
Let's actually keep it `var` - since our configs is not yet set up to approve these, it's throwing errors on build and in the CI.

georgestephanis replied:
Swapping back, but our precommit checks are conflicting with some of our configs in ci it seems --

![screen shot 2018-10-30 at 10 28 37 am](https://user-images.githubusercontent.com/941023/47725280-a6cf2600-dc2e-11e8-885f-f50e6acdd0a5.png)

dereksmart replied:
yeah, that's annoying and we should fix that inconsistency. For now, you can add `/* eslint-disable no-var */` to the files.

jeherve replied:
That's a known issue, you can use `--no-verify` here, Travis does not run the same tests.

More details about this:
p1540478539000100-slack-jetpack-crew

dereksmart commented on modules/calypsoify/class.jetpack-calypsoify.php:
```diff
@@ -38,11 +38,134 @@ public function setup() {
+
+			<span class="form-toggle__wrapper">
+				<input
+						id="autoupdate_plugin-toggle-<?php echo esc_attr( $slug ) ?>"
```
Supersized indents!

dereksmart commented on modules/calypsoify/class.jetpack-calypsoify.php:
```diff
@@ -38,11 +38,134 @@ public function setup() {
+			$list = Jetpack_Options::get_option( 'autoupdate_plugins', array() );
+			$initial_qty = sizeof( $list );
+
+			if ( $action === 'jetpack_enable_plugin_autoupdates' ) {
```
yoda-ize

dereksmart reviewed (dismissed):
Just a few minor nitpicks for now, with a couple of over-arching things we need to look at: 

1. I'm not seeing the success notices (via toggling). Doesn't look like anything in the .js to handle the `wp_send_json_success` sent from PHP yet, so probably just haven't gotten to it yet. 
2. We will need to add some logic to disable the toggles for plugins that are not in the repo (cannot set to auto-update)

Over all, looking good!

georgestephanis committed:
Only allow autoupdating plugins that are available in the plugins repo.

georgestephanis committed:
Swap back from let/const to var for ci compat

georgestephanis committed:
yoda soda

georgestephanis committed:
how the devil did that whitespace get off

jeherve commented:
I believe this will now need a rebase because of #10368

georgestephanis committed:
Merge branch 'master' into iterate/calypsoify

georgestephanis committed:
This was an oops from a merge -- shouldn't have been in this pr.

dereksmart commented on modules/calypsoify/build-sass.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+sass style.scss | perl -pe 's/^    /\t\t/g' | perl -pe 's/^  /\t/g' > style.css
```
Let's add these to our existing build process

georgestephanis replied:
Sure.  For the moment this was faster and easier to just do in place.

georgestephanis replied:
Done.

georgestephanis committed:
Merge branch 'master' into iterate/calypsoify

georgestephanis committed:
Roll in changes from master and convert them to the scss files.

Includes:

303a55c7de1e250881a0e6d75bc871466dd50af4
345c65e5560cc32ce8b53fe8b897f5f878fda324
472d2381886d79de392c0a652040f349039f69bf
1547d00f5fbcf804b1e04b1449810ce315afe30d
393042acd6514f959342523fd3dc6ca4959ba4a2

georgestephanis committed:
Add build tools.

georgestephanis committed:
Switch to minified, built versions of the assets.

jeherve commented on modules/calypsoify/class.jetpack-calypsoify.php:
```diff
@@ -90,17 +227,25 @@ public function add_plugin_menus() {
 
 	public function enqueue() {
-		wp_enqueue_style( 'calypsoify_wpadminmods_css', plugin_dir_url( __FILE__ ) . 'calypsoify/style.css', false, JETPACK__VERSION );
+		wp_enqueue_style( 'calypsoify_wpadminmods_css', plugin_dir_url( __FILE__ ) . 'style.min.css', false, JETPACK__VERSION );
```
What would you say about using `Jetpack::get_file_url_for_environment` for those minified files, so we can get both the minified or full versions when needed?

georgestephanis replied:
From what I saw it didn't seem as though the build tools were generating unminified versions of the non-rtl css files?  I'll recheck to be sure.

georgestephanis replied:
Yeah, it's not generating unminified versions of the css files -- there's just the `scss` and `-min`

jeherve reviewed (commented):
A quick note. I also wonder if we should leave a placeholder in place of `class.jetpack-calypsoify.php`, like we usually do to avoid issues (e.g. https://github.com/Automattic/jetpack/blob/master/modules/debug.php#L3 )?

georgestephanis commented:
The placeholder isn't necessary as it was only included by a hardcoded file, not a dynamic one based on database settings like modules are managed via -- so there shouldn't be any instances where the old file is attempted to be included.

georgestephanis committed:
Tidy up doRTL: to account for more than just two options.

georgestephanis committed:
Built versions of Calypsoify styles

rickybanister commented:
How can I test this with a plugin that isn't eligible for auto-updates? Something like the woo-beta-tester plugin? I was unable to build a site that both used this branch and had that plugin pre-installed. Whenever I added the url parameter for calypsoify I saw the traditional styles. 

My goal is to confirm how the toggle UI looks when auto-updates are not available. Everything else looks good.

georgestephanis committed:
Oops, use the right path for min files.

rickybanister commented:
I tested this out, everything functions as expected, even when auto-updates are not available (like for the Jetpack beta plugin). 

The only thing I noticed was the toggle's off state doesn't match the Calypso toggle's color:

https://wpcalypso.wordpress.com/devdocs/design/form-fields

I believe the color should be `background: $gray-lighten-10;` which compiles to #a8bece

dereksmart reviewed (dismissed):
Works well! Tried to break as much as I could. Let's update those colors that Rick pointed out and ship it!

georgestephanis committed:
Tweak color on toggle to match calypso

Co-Authored-By: Rick Banister <rickybanister@users.noreply.github.com>

rickybanister commented:
I just tried again (using the short-lived site link above) and didn't see a change. Does that link need to be updated or anything?

![image](https://user-images.githubusercontent.com/437258/48425645-41515e00-e733-11e8-9c58-93bf0678b7d1.png)

georgestephanis committed:
Correcting colors.

Now that I got the right selector, we can get the right one changed.

georgestephanis commented:
THERE WE GO, should work now.

rickybanister commented:
I just tried again using the same short-lived link from the description and there's no change to the color.

georgestephanis commented:
It's changed in the selector that you indicated, not sure why it's not showing for you -- are you sure that's what you'e looking at?

<img width="1041" alt="screen shot 2018-11-14 at 2 53 47 pm" src="https://user-images.githubusercontent.com/941023/48511752-3d9a0600-e81d-11e8-8fc1-04eb72ba7a9c.png">

dereksmart commented:
@rickybanister Something must have got stuck again in the beta plugin :( I'm seeing the correct color so I'll approve on your behalf and assuming there's no other blocking design feedback. 

![image](https://user-images.githubusercontent.com/7129409/48512382-f90f6a00-e81e-11e8-8433-229a5aa8901c.png)

rickybanister commented:
Perfect, thanks @georgestephanis and @dereksmart, sorry I didn't bother to check the branch manually in the beta plugin.

Ship it.

dereksmart reviewed (approved):
let's do it

jeherve commented:
Cherry-picked to `branch-6.7` in d760501e4af8f7947fcae0cdbd35a7fe9820a01b
